consts/src/lib.rs

```diff
@@ -91,36 +92,47 @@ mod structs {
     pub type BytesMac = [u8; MAC_LENGTH];
     pub type BytesEncodedVoucher = [u8; ENCODED_VOUCHER_LEN];
 
-    #[repr(C)]
-    #[derive(Default, PartialEq, Copy, Clone, Debug)]
-    pub enum EDHOCState {
-        #[default]
-        Start = 0, // initiator and responder
-        WaitMessage2 = 1,      // initiator
-        ProcessedMessage2 = 2, // initiator
-        ProcessedMessage1 = 3, // responder
-        WaitMessage3 = 4,      // responder
-        Completed = 5,         // initiator and responder
-    }
+    // This is sealed
+    pub trait EDHOCState: core::fmt::Debug {}
+    // For both initiator and responder
+    #[derive(Debug)]
+    pub struct Start;
+    impl EDHOCState for Start {}
+    // For the initiator
+    #[derive(Debug)]
+    pub struct WaitMessage2;
+    impl EDHOCState for WaitMessage2 {}
+    #[derive(Debug)]
+    pub struct ProcessedMessage2;
+    impl EDHOCState for ProcessedMessage2 {}
+    // For the responder
+    #[derive(Debug)]
+    pub struct ProcessedMessage1;
+    impl EDHOCState for ProcessedMessage1 {}
+    #[derive(Debug)]
+    pub struct WaitMessage3;
+    impl EDHOCState for WaitMessage3 {}
+    // For both again
+    #[derive(Debug)]
+    pub struct Completed;
+    impl EDHOCState for Completed {}
 
     #[repr(C)]
     #[derive(PartialEq, Debug)]
     pub enum EDHOCError {
-        Success = 0,
         UnknownPeer = 1,
         MacVerificationFailed = 2,
         UnsupportedMethod = 3,
         UnsupportedCipherSuite = 4,
         ParsingError = 5,
-        WrongState = 6,
         EADError = 7,
         UnknownError = 8,
     }
 
     #[repr(C)]
-    #[derive(Default, Copy, Clone, Debug)]
-    pub struct State(
-        pub EDHOCState,
+    #[derive(Debug)]
+    pub struct State<Phase: EDHOCState>(
+        pub PhantomData<Phase>,
```

Comment:
Yes. It's a zero-sized type that is necessary to justify toward Rust that the types are different, and while it's often put at the end, here I'm putting it in the very position that used to inhabit the runtime state, because it means that the tuple struct unpacking doesn't need to change.